### PR TITLE
Changes to Metadb derived table po_lines_locations.sql with comments added

### PR DESCRIPTION
**Pull request by natalyapik**
Added holding_id and relying locations resolving on newly introduced holding information at po line level for derived table po_lines_locations (Metadb version)
```
All queries:
- [ ] Release notes added or updated in CHANGES.md
- [ ] Query runs without errors
- [ ] Query output is correct
- [ ] Query logic is clear and well documented
- [ ] Query is readable and properly indented with spaces (not tabs)
- [ ] Table and column names are in all-lowercase
- [ ] Quotation marks are used only where necessary

Report queries:
- [ ] Query has complete user documentation
    - [ ] Purpose of report
    - [ ] Sample output
    - [ ] Query instructions

Derived tables:
- [ ] Query begins with user documentation in comment lines
- [ ] File name is listed in `runlist.txt` after dependencies
- [ ] All columns have indexes
- [ ] Table is vacuumed and analyzed
```

### CODE REVIEW

**Review by amzoss — CHANGES_REQUESTED** [review]
I know this query is already in production and you're just making a few small changes, but while we're making edits, I think this could be improved to meet our current standards. I have a few suggestions (below the checklist).

All queries:
- [x] Release notes added or updated in CHANGES.md
- [x] Query runs without errors
- [x] Query output is correct
- [ ] Query logic is clear and well documented
- [ ] Query is readable and properly indented with spaces (not tabs)
- [x] Table and column names are in all-lowercase
- [x] Quotation marks are used only where necessary

Derived tables:
- [x] Query begins with user documentation in comment lines
- [x] File name is listed in `runlist.txt` after dependencies
- [x] All columns have indexes
- [x] Table is vacuumed and analyzed

This query is doing some complicated logic because of the weird data model. I think there are ways that this could be more clear. You could try one or more of the following:

* instead of `loc` and `loc2`, the joined locations tables could be named more clearly based on the field they're connecting to, like `pol_location` and `holdings_location`
* the reason for the `CASE` statements could be explained in the documentation at the top or in short comments embedded within the query; understanding why these fields work the way they do is important to using the data, and I think it might be nice to have that spelled out right in the query 
* I think the indenting for the `CASE` statements could be a bit better to help signal the logic a bit better. Here's how I would do it:

```
    CASE 
        WHEN ploc.pol_location_id IS NOT NULL THEN ploc.pol_location_id  
        ELSE hr.permanent_location_id
    END AS pol_location_id, 
    CASE 
        WHEN loc.name IS NOT NULL THEN loc.name
        ELSE loc2.name
    END AS pol_location_name,
    CASE 
        WHEN loc.name IS NOT NULL THEN 'pol_location'
        WHEN loc2.name IS NOT NULL THEN 'pol_holding'
        ELSE 'no_source'
    END AS pol_location_source
```

Also, the file name does not match the name of the derived table. Is that something we should be worried about? (I assume we would change the file name and not the table name.)

Finally there are a few typos I've noticed. I'll comment on the lines directly.

File: sql_metadb/derived_tables/po_lines_location.sql [hunk 1]
```diff
@@ -17,25 +18,51 @@ WITH ploc AS (
...
+    CASE WHEN loc.name IS NOT NULL THEN 'pol_location'
+         WHEN loc2.name IS NOT NULL THEN 'pol_holding'
+         ELSE 'no_source'
+    END AS pol_location_source
 FROM
     ploc
-    LEFT JOIN folio_inventory.location__t AS loc ON loc.id = ploc.pol_loc_id;
+    LEFT JOIN folio_inventory.holdings_record__t AS hr ON ploc.pol_holding_id::uuid= hr.id
+    LEFT JOIN folio_inventory.location__t AS loc ON loc.id = ploc.pol_location_id
+    LEFT JOIN folio_inventory.location__t AS loc2 ON loc2.id ::uuid= hr.permanent_location_id ;
```
Comment (amzoss) [inline sql_metadb/derived_tables/po_lines_location.sql:38]: There is a space between `loc2.id` and `::uuid`. Not sure if this is a problem, but it's inconsistent. Maybe there could also be a space added between `uuid` and `=`.

File: sql_metadb/derived_tables/po_lines_location.sql [hunk 2]
```diff
@@ -17,25 +18,51 @@ WITH ploc AS (
...
+         ELSE loc2.name
+    END AS pol_location_name,
+    CASE WHEN loc.name IS NOT NULL THEN 'pol_location'
+         WHEN loc2.name IS NOT NULL THEN 'pol_holding'
+         ELSE 'no_source'
+    END AS pol_location_source
 FROM
     ploc
-    LEFT JOIN folio_inventory.location__t AS loc ON loc.id = ploc.pol_loc_id;
+    LEFT JOIN folio_inventory.holdings_record__t AS hr ON ploc.pol_holding_id::uuid= hr.id
```
Comment (amzoss) [inline sql_metadb/derived_tables/po_lines_location.sql:36]: Maybe there could be a space added between `uuid` and `=`.

File: sql_metadb/derived_tables/po_lines_location.sql [hunk 3]
```diff
@@ -17,25 +18,51 @@ WITH ploc AS (
...
-CREATE INDEX ON po_lines_locations (location_name);
+CREATE INDEX ON po_lines_locations (pol_location_id);
+
+CREATE INDEX ON po_lines_locations (pol_location_name);
+
+CREATE INDEX ON po_lines_locations (pol_location_source);
+
+COMMENT ON COLUMN po_lines_locations.pol_id IS 'UUID identifying this purchase order line';
+
+COMMENT ON COLUMN po_lines_locations.pol_loc_qty IS 'combined/total quanitity of physical and electronic items';
```
Comment (amzoss) [inline sql_metadb/derived_tables/po_lines_location.sql:56]: `quanitity` should be `quantity`

File: sql_metadb/derived_tables/po_lines_location.sql [hunk 4]
```diff
@@ -17,25 +18,51 @@ WITH ploc AS (
...
+
+COMMENT ON COLUMN po_lines_locations.pol_loc_qty_elec IS 'Quantity of electronic items in this purchase order line';
+
+COMMENT ON COLUMN po_lines_locations.pol_loc_qty_phys IS 'Quantity of physical items or resources of "Other" order format in this purchase order line';
+
+COMMENT ON COLUMN po_lines_locations.pol_location_id IS 'UUID of the (inventory) location record or Holding UUID associated with order line';
+
+COMMENT ON COLUMN po_lines_locations.pol_location_name IS 'Name of the location assosiated with pol_location_id';
```
Comment (amzoss) [inline sql_metadb/derived_tables/po_lines_location.sql:64]: `assosiated` should be `associated`

File: sql_metadb/derived_tables/po_lines_location.sql [hunk 5]
```diff
@@ -17,25 +18,51 @@ WITH ploc AS (
...
+
+COMMENT ON COLUMN po_lines_locations.pol_location_source IS 'Source of the location assosiated with pol_location_id';
```
Comment (amzoss) [inline sql_metadb/derived_tables/po_lines_location.sql:66]: `assosiated` should be `associated`

**Comment by amzoss** [comment]
@natalyapik I requested a few changes for this PR. I know they're mostly just typos, but would you be willing to go ahead and propose some updates? I think making the logic easier to understand would be a pretty high priority for me, since that's a major change to this query and it could be difficult to follow. Fixing the typos is not a high priority, so if that's holding you back we could skip those suggestions.

File: sql_metadb/derived_tables/po_lines_location.sql (hunk 4 above)
Comment (sbeltaine) [reply sql_metadb/derived_tables/po_lines_location.sql:64]: Could you clarify what still needs to be address on this PR?

Comment (amzoss) [reply sql_metadb/derived_tables/po_lines_location.sql:64]: I suggested many changes in my initial review. 

https://github.com/folio-org/folio-analytics/pull/697#pullrequestreview-1201801632

My main concern is updating the query to help explain/document the logic changes better. I also identified line by line several typos and style inconsistencies that could be updated at the same time.

File: sql_metadb/derived_tables/po_lines_location.sql [hunk 6]
```diff
@@ -17,25 +18,51 @@ WITH ploc AS (
...
+
+COMMENT ON COLUMN po_lines_locations.pol_id IS 'UUID identifying this purchase order line';
+
+COMMENT ON COLUMN po_lines_locations.pol_loc_qty IS 'combined/total quanitity of physical and electronic items';
+
+COMMENT ON COLUMN po_lines_locations.pol_loc_qty_elec IS 'Quantity of electronic items in this purchase order line';
+
+COMMENT ON COLUMN po_lines_locations.pol_loc_qty_phys IS 'Quantity of physical items or resources of "Other" order format in this purchase order line';
+
+COMMENT ON COLUMN po_lines_locations.pol_location_id IS 'UUID of the (inventory) location record or Holding UUID assosiated with order line';
```
Comment (amzoss) [inline sql_metadb/derived_tables/po_lines_location.sql:62]: Again, typo with `assosiated`. The correct spelling is `associated`.

**Review by sbeltaine — APPROVED** [review]
Natalya has made the requested changes. Approved.

**Review by amzoss — APPROVED** [review]
Excellent, thanks for the changes! This looks great to me.

All queries:
- [x] Release notes added or updated in CHANGES.md
- [x] Query runs without errors
- [x] Query output is correct
- [x] Query logic is clear and well documented
- [x] Query is readable and properly indented with spaces (not tabs)
- [x] Table and column names are in all-lowercase
- [x] Quotation marks are used only where necessary

Derived tables:
- [x] Query begins with user documentation in comment lines
- [x] File name is listed in `runlist.txt` after dependencies
- [x] All columns have indexes
- [x] Table is vacuumed and analyzed

**Pull request merged**